proofs/tree.go

```diff
@@ -182,47 +179,50 @@ import (
 	"github.com/centrifuge/precise-proofs/proofs/proto"
 	"github.com/golang/protobuf/proto"
 	"github.com/xsleonard/go-merkle"
+	"github.com/pkg/errors"
 )
 
-// DefaultSaltsLengthSuffix is the suffix used to store the length of slices (repeated) fields in the tree. It can be
-// customized with the SaltsLenghtSuffix TreeOption
-const DefaultSaltsLengthSuffix = "length"
+// DefaultReadablePropertyLengthSuffix is the suffix used to store the length of slices (repeated) fields in the tree. It can be
+// customized with the ReadablePropertyLengthSuffix TreeOption
+const DefaultReadablePropertyLengthSuffix = "length"
 
 // TreeOptions allows customizing the generation of the tree
 type TreeOptions struct {
 	//	EnableHashSorting: Implement a merkle tree with sorted hashes
 	EnableHashSorting bool
-	GetSalt           GetSalt
-	Salts             *Salts
-	// SaltsLengthSuffix: As precise proofs support repeated fields, when generating the merkle tree we need to add a
-	// leaf that represents the length of the slice. The default suffix is `Length`, although it is customizable so it
+	Salts             Salts
+	// ReadablePropertyLengthSuffix: As precise proofs support repeated fields, when generating the merkle tree we need to add a
+	// leaf that represents the length of the slice. The default suffix is `_length`, although it is customizable so it
 	// does not collide with potential field names of your own proto structs.
-	SaltsLengthSuffix string
+	ReadablePropertyLengthSuffix string
 	Hash              hash.Hash
 	// ParentPrefix defines an arbitrary prefix to prepend to the parent, so all fields are prepended with it
 	ParentPrefix      Property
 	CompactProperties bool
 }
 
-type Salt struct {
-	Compact []byte
-	Value   []byte
-}
-type Salts []Salt
-type GetSalt func(compact []byte) []byte
+type Salts func(compact []byte) ([]byte, error)
 
-func defaultGetSalt(salts *Salts) func([]byte) []byte {
-	return func(compact []byte) []byte {
-		for ii := range *salts {
-			if bytes.Compare((*salts)[ii].Compact, compact) == 0 {
-				return (*salts)[ii].Value
+func defaultGetSalt(message proto.Message) (Salts, error) {
+	salts, err := getSaltsFromMessage(message)
+	if (err != nil) {
+		return nil, err
+	}
+	return func(compact []byte) ([]byte,error) {
+		for _, salt := range salts {
+			if bytes.Compare(salt.GetCompact(), compact) == 0 {
+				return salt.GetValue(), nil
 			}
 		}
 		randbytes := make([]byte, 32)
-		rand.Read(randbytes)
-		*salts = append(*salts, Salt{Compact: compact, Value: randbytes})
-		return randbytes
-	}
+		n, err := rand.Read(randbytes)
+		if (err != nil){
+			return nil, err
+		} else if (n != 32) {
+			return nil, errors.Wrapf(err, "Only read %d instead of 32 random bytes", n)
+		}
```

Comment:
The generated salt `randbytes` here needs to be persisted in the salts on the protobuf message. Otherwise it would be impossible to reconstruct the same tree because the salt would change every time.

Reply:
@lucasvo  
You are right, but currently I cannot return generated salts by embeded salts field because all messages defined in  example.proto does not contain `repeated salts Salt`，
I can change `func (doctree *DocumentTree) AddLeavesFromDocument(document proto.Message) (err error)` to  `func (doctree *DocumentTree) AddLeavesFromDocument(document proto.Message) (salts []*proofspb.Salt, err error)`
Do you think this solution is ok for you?

Reply:
Would you be able to add the salts to the ExampleMessage proto and adjust the testcase for it?